[WIP] 511 implementation using a service object

elg0nz opened the pull request:
* [x] Extract functionality into a Service object
* [x] Add test coverage by porting existing view tests to the Service Object tests
* [x] Test all the scenarios on the Tock #594 
* [x] Profit.

elg0nz committed:
Fix #511 Implement AddHours endpoint

elg0nz committed:
Introduce HoursAdder service object

gotascii commented on tock/hours/views.py:
```diff
@@ -313,6 +315,29 @@ def form_valid(self, form):
+    reporting_period = ReportingPeriod.objects.latest()
+    hours_adder = HoursAdder(
+        project_id = request.GET['project'],
+        hours = Decimal(request.GET['hours']),
```
The Decimal-ing of `request.GET['hours']` should probs be handled in `HoursAdder` as it is a failure case we need to support (e.g. Someone submits the value `hi` for `request.GET['hours']`)

gotascii commented on tock/hours/hours_adder.py:
```diff
@@ -0,0 +1,70 @@
+
+        updated_hours = tco.hours_spent + self.hours
+        updated_hours = max(0, updated_hours)
+        tco.hours_spent = updated_hours
```
This was my doing, but I think the above 2 lines could be combined w/o loosing clarity: `tco.hours_spent = max(0, updated_hours)`.

elg0nz replied:
👍

gotascii commented on tock/hours/hours_adder.py:
```diff
@@ -0,0 +1,70 @@
+        self.user_id = kwargs.get('user_id', None)
+        self.operation_was_successful = False
+
+    def generate_successful_message(self, hours, project):
```
I feel like all this should maybe stay in the controller level (perhaps as another class) as it deals with creating urls and html. HoursAdder could just report success or failure and why in a non-web-based way. The controller should probably be in charge of then generating the url and html.

elg0nz committed:
Handle Decimal parsing on HoursAdder

elg0nz committed:
Make the controller provide the undo_url for HoursAdder

elg0nz committed:
Port add_hours tests from it's view to test_hours_adder

elg0nz committed:
Fix bug while comparing dates on TimeCardTests

Per the Python docs
https://docs.python.org/2/library/datetime.html#datetime.datetime.today

datetime.datetime.today() returns current local datetime with tzinfo
None.

Since the timecard objects during testing are stored on UTC time
(according to my local tests), the tests fail when the datetime.date
drifts enough from utc to return a different day.

This updates the tests to use datetime.datetime.utcnow().day instead,
which prevents this issue.

We also recommend evaluating  https://github.com/spulec/freezegun to make
these date comparisons deterministic.

elg0nz committed:
Also test undo on test_views.py#test_add_hours

elg0nz committed:
Add error handling for invalid inputs to HoursAdder

elg0nz committed:
Prevent negative hours to be created via HoursAdder
